cdap-unit-test/src/test/java/co/cask/cdap/test/app/ServiceLifeCycleTestRun.java

```diff
@@ -256,6 +275,142 @@ public Long call() throws Exception {
+
+    } finally {
+      System.clearProperty(ServiceHttpServer.THREAD_POOL_SIZE);
+    }
+  }
+
+  @Test
+  public void testContentConsumerProducerLifecycle() throws Exception {
+    // Set to have one thread only for testing context capture and release
+    System.setProperty(ServiceHttpServer.THREAD_POOL_SIZE, "1");
+
+    try {
+      ApplicationManager appManager = deployApplication(ServiceLifecycleApp.class);
+      final ServiceManager serviceManager = appManager.getServiceManager("test").start();
+      final DataSetManager<KeyValueTable> datasetManager = getDataset(ServiceLifecycleApp.HANDLER_TABLE_NAME);
+      // Clean up the dataset first to avoid being affected by other tests
+      datasetManager.get().delete(Bytes.toBytes("called"));
+      datasetManager.get().delete(Bytes.toBytes("completed"));
+      datasetManager.flush();
+
+      CountDownLatch uploadLatch = new CountDownLatch(1);
+
+      // Create five concurrent upload
+      List<ListenableFuture<Integer>> completions = new ArrayList<>();
+      for (int i = 0; i < 5; i++) {
+        completions.add(slowUpload(serviceManager, "POST", "uploadDownload", uploadLatch));
+      }
+
+      // Get the states, there should be six handler instances initialized.
+      // Five for the in-progress upload, one for the getStates call
+      Tasks.waitFor(6, new Callable<Integer>() {
+        @Override
+        public Integer call() throws Exception {
+          return getStates(serviceManager).size();
+        }
+      }, 5, TimeUnit.SECONDS, 100, TimeUnit.MILLISECONDS);
+
+      // Complete the upload
+      uploadLatch.countDown();
+
+      // Make sure the download through content producer has started
+      Tasks.waitFor(true, new Callable<Boolean>() {
+        @Override
+        public Boolean call() throws Exception {
+          byte[] value = datasetManager.get().read("called");
+          datasetManager.flush();
+          if (value == null || value.length != Bytes.SIZEOF_LONG) {
+            return false;
+          }
+          return Bytes.toLong(value) > 5;
+        }
+      }, 10L , TimeUnit.SECONDS, 100, TimeUnit.MILLISECONDS);
+
+      // Get the states, there should still be six handler instances since the ContentConsumer should
+      // be passing it's captured context to the ContentProducer without creating new one.
+      Multimap<Integer, String> states = getStates(serviceManager);
+      Assert.assertEquals(6, states.size());
+
+      // Set the complete flag in the dataset
+      datasetManager.get().write("completed", Bytes.toBytes(true));
+      datasetManager.flush();
+
+      // Wait for completion
+      Futures.successfulAsList(completions).get(10, TimeUnit.SECONDS);
+
+      // Verify the upload result
+      for (ListenableFuture<Integer> future : completions) {
+        Assert.assertEquals(200, future.get().intValue());
+      }
+
+      // Get the states again, it should still be 6 same instances
```

Comment:
can they expire before you make this call, on a slow machine? That would lead to a flaky test...